### ISSUE TITLE
ProjectionPlot throws a MPL error when attempting to save()

### ISSUE THREAD

**Issue by chummels** [yt-project/yt#2890]
### Bug report

**Bug summary**

`ProjectionPlot` throwing an Matplotlib error on yt sample dataset, when it never did in the past.  This is part of a script in the standard tests used in Trident, which has long worked OK.  Note, that this simple example doesn't require trident at all in order to run.  Perhaps this is related to the transition to the most recent version of matplotlib?

**Code for reproduction**

```python
import yt
fn = 'FIRE_M12i_ref11/snapshot_600.hdf5'
ds = yt.load(fn)
p = yt.ProjectionPlot(ds, 'x', ('gas', 'density'))
p.save('projection.png')
```

**Actual outcome**

```
Traceback (most recent call last):
  File "gizmo_script.py", line 5, in <module>
    p.save('projection.png')
  File "/Users/chummels/src/yt/yt/visualization/plot_container.py", line 79, in newfunc
    rv = f(*args, **kwargs)
  File "/Users/chummels/src/yt/yt/visualization/plot_container.py", line 505, in save
    names.append(v.save(name, mpl_kwargs))
  File "/Users/chummels/src/yt/yt/visualization/base_plot_types.py", line 162, in save
    canvas.print_figure(name, **mpl_kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/backend_bases.py", line 2217, in print_figure
    **kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/backend_bases.py", line 1639, in wrapper
    return func(*args, **kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 509, in print_png
    FigureCanvasAgg.draw(self)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 407, in draw
    self.figure.draw(self.renderer)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/figure.py", line 1864, in draw
    renderer, self, artists, self.suppressComposite)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/image.py", line 131, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/cbook/deprecation.py", line 411, in wrapper
    return func(*inner_args, **inner_kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 2748, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/image.py", line 131, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/artist.py", line 41, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/image.py", line 638, in draw
    renderer, renderer.get_image_magnification())
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/image.py", line 923, in make_image
    magnification, unsampled=unsampled)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/image.py", line 541, in _make_image
    output = self.norm(resampled_masked)
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/colors.py", line 1192, in __call__
    self._check_vmin_vmax()
  File "/Users/chummels/src/miniconda3/lib/python3.7/site-packages/matplotlib/colors.py", line 1181, in _check_vmin_vmax
    raise ValueError("minvalue must be positive")
ValueError: minvalue must be positive
```

**Expected outcome**

Generate a projection.png image of the ProjectionPlot.

**Version Information**

  * Operating System: OS X 15.6
  * Python Version: Python 3.7.7
  * yt version: `master` tip from yt repository as of 8/28/20: (hash e72fc9a31518)
  * Other Libraries (if applicable): matplotlib 3.3.1

**Comment by chummels** [comment]
When I switch the plot to linear scaling, it shows that the bulk of the space is at 0 column density value, which I guess is causing the failure.  Although I imagine this should've been the case before too.  When I have a second, I'll try going back to a previous version of the code to see what it looked like.  Perhaps something got toggled with the kernel or "scatter" settings for things that caused part of the space to be identically 0 instead of empty/NaN?  Just speculation.

![Linear Scaling](https://user-images.githubusercontent.com/8250999/91625563-3cd45480-e95d-11ea-8ee0-6181faf931f8.png)

**Comment by chrishavlin** [comment]
I **very briefly** looked into this -- a superficial fix would be to adjust the check for whether or not to use the symlog scaling. Right now it checks

```python
elif np.nanmax(image) > 0.0 and np.nanmin(image) < 0
```

in `PWViewerMPL._setup_plots()`. Switching this to

```python
elif np.nanmax(image) > 0.0 and np.nanmin(image) <= 0
```

would switch to symlog in the present case and the matplotlib error would not get thrown. I also frequently work with fields that are easier to visualize in log-space but can have 0 values (I often do some masking or set a nonzero min value), so I would find this helpful. But I don't know what has changed to cause the error now, so even if others agreed that this symlog change was appropriate, I don't think it should go in until the reason for the change in behavior is figured out...

**Comment by jzuhone** [comment]
This started after matplotlib 3.3.x, where I am not sure what the "x" is, so it's probably a change on their end. I think they started throwing an error if zeros were in log-plots like this.

**Comment by matthewturk** [comment]
This seems like a reasonable fix.

**Comment by chummels** [comment]
Sounds good to me!

**Comment by jzuhone** [comment]
@chummels for this file, at what scale should I expect structure?

**Comment by chrishavlin** [comment]
Moving discussion back here since it seems more appropriate.... 

So I think we're dealing with a rounding error matplotlib bug here. When you set a log-scale using `matplotlib.colors.LogNorm` without setting `vmin` and `vmax` (as yt does), `LogNorm` will calculate `vmin` and `vmax` using a masked array that excludes zero values. For this projection, this gives `vmin = 4.8845906468177196e-23` and `vmax = 3.549929801742279`.  This is all well and good until the final image gets drawn... in `matplotlib.image._ImageBase._make_image()` there is some complicated rescaling that does stuff like `dv = np.float64(self.norm.vmax) - np.float64(self.norm.vmin)` and addition/subtraction of offset values. In the case of these `vmin` and `vmax` values, after all that `vmin` ends up set to 0, and we end up with our error here. So that does seem like a matplotlib bug... I'll try to work out a more minimal example to reproduce this and submit a bug report via matplotlib. 

That said, the following snippet shows that the projection plot DOES work as expected with weight field that produces a more reasonable (?) data range:

```python
import yt
fn = 'FIRE_M12i_ref11/snapshot_600.hdf5'
ds = yt.load(fn)
p = yt.ProjectionPlot(ds, 'x', ('gas', 'density'), weight_field=('gas', 'density'))
p.show()
```
![proj_self](https://user-images.githubusercontent.com/22038879/113359009-7c588a00-930c-11eb-8f92-acdd212defdf.png)

So perhaps there is a yt solution here related to the how the `weight_field` is set (perhaps related to PR #2211) ?

**Comment by jzuhone** [comment]
@chrishavlin nice work sorting this out!

Is there something we can do even when there isn't a `weight_field`? I'd rather not wait on a Matplotlib release if we can avoid it somehow.

**Comment by chrishavlin** [comment]
Yes! I'm sure we can find a fix that doesn't require relying on a matplotlib release -- not yet sure where that fix should go in.

**Comment by chrishavlin** [comment]
with a little more digging on the matplotlib side, I found this issue https://github.com/matplotlib/matplotlib/issues/18415 the fix that went in was not general enough. 

In terms of a fix on the yt side, I can see a couple approaches:

1. just use the existing symlog fix and some more comments (would require updating the failing image tests?)
2. check the range of the data and if it's greater than some cutoff value and there are 0's, use symlog. (e.g., `if np.log10(vmax) - np.log10(vmin) > cutoff and the image has zeros`, use symlog)
3. use a linear colorscale but plot `np.log10(the_data)`?

**Comment by munkm** [comment]
Closed via yt-project/yt#3161!!!
